File: week-3/edit-distance-recursive-memoized/edit-distance-recursive-memoized.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {number}
 */
const memo = {};

var minDistance = function(word1, word2) {
    const state = word1 + '~~' + word2;
    
    if (state in memo) return memo[state];
    if (word1.length === 0 && word2.length === 0) memo[state] = 0;
    else if (word1.length === 0) memo[state] = 1 + minDistance(word1, word2.slice(1));
    else if (word2.length === 0) memo[state] =  1 + minDistance(word1.slice(1), word2);
    else if (word1[0] === word2[0]) memo[state] = minDistance(word1.slice(1), word2.slice(1));
    
    else memo[state] =  1 + Math.min(minDistance(word1.slice(1), word2.slice(1)), minDistance(word1.slice(1), word2), minDistance(word1, word2.slice(1)));

    return memo[state];
};

File: week-3/edit-distance-recursive-memoized/edit-distance-recursive-memoized.py
class Solution:
    def __init__(self):
        self.memo = {}
        
    def minDistance(self, word1: str, word2: str) -> int:
        state = word1 + '~~' + word2
        if state in self.memo:
            return self.memo[state]
        
        elif len(word1) == 0 and len(word2) == 0:
            self.memo[state] = 0
        
        elif len(word1) == 0:
            self.memo[state] = 1 + self.minDistance(word1, word2[1:])
        
        elif len(word2) == 0:
            self.memo[state] = 1 + self.minDistance(word1[1:], word2)
            return self.memo[state]
        
        elif word1[0] == word2[0]:
            self.memo[state] = self.minDistance(word1[1:], word2[1:])
        else:
            self.memo[state] =  1 + min(self.minDistance(word1[1:], word2[1:]), self.minDistance(word1[1:], word2), self.minDistance(word1, word2[1:]))
            
        return self.memo[state]

